File: src/NextPermutation.java
import java.util.Arrays;

/**
 * User: guoxc
 * Date: 15-1-17
 * Time: 14:05
 */
public class NextPermutation {
    public void nextPermutation(int[] num){
        int j = -1;
        for(int i = num.length-2; i > -1; i -= 1){
            if(num[i] < num[i+1]){
                j = i;
                break;
            }
        }
        if(j == -1){
            reverse(num, 0, num.length);
            return;
        }
        int min = num[j+1];
        int k = j + 1;
        for(int i = j + 1; i < num.length; i += 1){
            if(num[i] > num[j] && num[i] <= min){
                min = num[i];
                k = i;
            }
        }
        swap(num, j, k);
        reverse(num, j + 1, num.length);
        return;

    }
    public void swap(int[] num, int i, int j){
        int temp = num[j];
        num[j] = num[i];
        num[i] = temp;
    }
    public void reverse(int[] num, int i, int j){
        int l = j - i;
        if(l < 2) return;
        for(int k = i; k < l / 2 + i; k += 1){
            swap(num, k, j - 1 - k + i);
        }
        return;
    }
    public static void test(){
        NextPermutation instance = new NextPermutation();
        int[] x1 = new int[]{2,3,1,3,3};
        instance.nextPermutation(x1);
        System.out.println(Arrays.toString(x1));
        int[] x2 = new int[]{3,2,1};
        instance.nextPermutation(x2);
        System.out.println(Arrays.toString(x2));
        int[] x3 = new int[]{1,1,5};
        instance.nextPermutation(x3);
        System.out.println(Arrays.toString(x3));
    }
}

File: cpp/next_permutation.cpp
class Solution {
private:
    void reverse(vector<int>& nums, int l, int r){
        while(l < r){
            swap(nums, l, r);
            l += 1;
            r -= 1;
        }
    }
    void swap(vector<int>& nums, int l, int r){
        int temp = nums[l];
        nums[l] = nums[r];
        nums[r] = temp;
    }
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int i = n - 1;
        while(i > 0 && nums[i] <= nums[i-1]){
            i -= 1;
        }
        if(i == 0){
            reverse(nums, 0, n - 1);
            return;
        }
        //nums[i - 1] < nums[i] > nums[i+1] > ...
        int min_n = INT_MAX;
        int min_i = -1;
        for(int j = n - 1; j >= i ; j -= 1){
            if(nums[j] < min_n && nums[j] > nums[i-1]){
                min_n = nums[j];
                min_i = j;
            }
        }
        if(min_i != -1){
            swap(nums, min_i, i - 1);
            reverse(nums, i, n - 1);
        }else{
            reverse(nums, i-1, n - 1);
        }
    }
};
